flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/windowing/windowbuffer/MedianPreReducer.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.windowing.windowbuffer;
+
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.streaming.api.windowing.StreamWindow;
+import org.apache.flink.streaming.api.windowing.helper.fieldaccessors.FieldAccessorCreator;
+import org.apache.flink.util.Collector;
+import org.apache.flink.streaming.api.windowing.helper.fieldaccessors.FieldAccessor;
+
+import java.util.TreeMap;
+import java.util.Map.Entry;
+import java.util.Deque;
+import java.util.ArrayDeque;
+import java.util.Comparator;
+import java.io.Serializable;
+
+/**
+ * Non-grouped pre-reducer for calculating median with any eviction policy.
+ */
+public class MedianPreReducer<T> extends WindowBuffer<T> implements PreAggregator, Serializable {
+
+	private static final long serialVersionUID = 1L;
+
+	// This is for getting and setting the field specified in the parameters to the median call.
+	private FieldAccessor<T, Double> fieldAccessor;
+
+	private TypeSerializer<T> serializer;
+
+	// These contain lower-than-median and higher-than-median elements.
+	// Elements that compare equal to the median can be in either.
+	// The invariant is that low always contains the same amount or one more element than high.
+	//   (see the assert in updateMedian)
+	TreeMultiset<T>
+			low = new TreeMultiset<T>(new CompareOnField()),
+			high = new TreeMultiset<T>(new CompareOnField());
+	Deque<T> elements = new ArrayDeque<T>();
+
+	T median;
+
+	public MedianPreReducer(int pos, TypeInformation<T> typeInfo, ExecutionConfig config) {
+		this.fieldAccessor = FieldAccessorCreator.create(pos, typeInfo);
+		this.serializer = typeInfo.createSerializer(config);
+	}
+
+	public MedianPreReducer(String field, TypeInformation<T> typeInfo, ExecutionConfig config) {
+		this.fieldAccessor = FieldAccessorCreator.create(field, typeInfo, config);
+		this.serializer = typeInfo.createSerializer(config);
+	}
+
+	private void updateMedian() {
+		assert low.size() == high.size() || low.size() == high.size() + 1;
```

Comment:
Assertions are not enabled in Flink.
I would suggest to use Guava's preconditions.